### ISSUE TITLE
mask模型的F1 Precision Recall IoU值计算

### ISSUE THREAD

**Issue by wuchaotao** [ultralytics/ultralytics#20243]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Yolo中的分割模型中，mask的F1 Precision Recall IoU值计算

**Comment by UltralyticsAssistant** [comment]
👋 Hello @wuchaotao, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response and an Ultralytics engineer will also assist you soon 😊

**Comment by glenn-jocher** [comment]
Hi @wuchaotao,

For mask segmentation models in Ultralytics YOLO, the calculation of metrics (F1, Precision, Recall, IoU) is handled similarly to bounding box detection, but operates on mask predictions instead.

The core calculation uses the `mask_iou` function in `ultralytics/utils/metrics.py`:

```python
def mask_iou(mask1, mask2, eps=1e-7):
    """Calculate masks IoU."""
    intersection = torch.matmul(mask1, mask2.T).clamp_(0)
    union = (mask1.sum(1)[:, None] + mask2.sum(1)[None]) - intersection
    return intersection / (union + eps)
```

This computes overlap between predicted and ground truth masks. Then metrics are calculated as:
- Precision: TP/(TP+FP) - ratio of correct mask predictions to all predictions
- Recall: TP/(TP+FN) - ratio of correct mask predictions to all ground truth masks
- F1 Score: 2×(Precision×Recall)/(Precision+Recall) - harmonic mean of precision and recall

The `ap_per_class` function processes these values across different IoU thresholds (commonly from 0.5 to 0.95) to compute the final performance metrics.

Would you like more specific details about any part of this calculation process?

**Comment by wuchaotao** [comment]
I get it, thank you so much

---- 回复的原邮件 ----
| 发件人 | Glenn ***@***.***> |
| 发送日期 | 2025年04月21日 09:54 |
| 收件人 | ultralytics/ultralytics ***@***.***> |
| 抄送人 | wuchaotao ***@***.***>,
Mention ***@***.***> |
| 主题 | Re: [ultralytics/ultralytics] mask模型的F1 Precision Recall IoU值计算 (Issue #20243) |

Hi @wuchaotao,

For mask segmentation models in Ultralytics YOLO, the calculation of metrics (F1, Precision, Recall, IoU) is handled similarly to bounding box detection, but operates on mask predictions instead.

The core calculation uses the mask_iou function in ultralytics/utils/metrics.py:

defmask_iou(mask1, mask2, eps=1e-7):
    """Calculate masks IoU."""intersection=torch.matmul(mask1, mask2.T).clamp_(0)
    union= (mask1.sum(1)[:, None] +mask2.sum(1)[None]) -intersectionreturnintersection/ (union+eps)

This computes overlap between predicted and ground truth masks. Then metrics are calculated as:

Precision: TP/(TP+FP) - ratio of correct mask predictions to all predictions
Recall: TP/(TP+FN) - ratio of correct mask predictions to all ground truth masks
F1 Score: 2×(Precision×Recall)/(Precision+Recall) - harmonic mean of precision and recall

The ap_per_class function processes these values across different IoU thresholds (commonly from 0.5 to 0.95) to compute the final performance metrics.

Would you like more specific details about any part of this calculation process?

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

glenn-jocher left a comment (ultralytics/ultralytics#20243)

Hi @wuchaotao,

For mask segmentation models in Ultralytics YOLO, the calculation of metrics (F1, Precision, Recall, IoU) is handled similarly to bounding box detection, but operates on mask predictions instead.

The core calculation uses the mask_iou function in ultralytics/utils/metrics.py:

defmask_iou(mask1, mask2, eps=1e-7):
    """Calculate masks IoU."""intersection=torch.matmul(mask1, mask2.T).clamp_(0)
    union= (mask1.sum(1)[:, None] +mask2.sum(1)[None]) -intersectionreturnintersection/ (union+eps)

This computes overlap between predicted and ground truth masks. Then metrics are calculated as:

Precision: TP/(TP+FP) - ratio of correct mask predictions to all predictions
Recall: TP/(TP+FN) - ratio of correct mask predictions to all ground truth masks
F1 Score: 2×(Precision×Recall)/(Precision+Recall) - harmonic mean of precision and recall

The ap_per_class function processes these values across different IoU thresholds (commonly from 0.5 to 0.95) to compute the final performance metrics.

Would you like more specific details about any part of this calculation process?

**Comment by glenn-jocher** [comment]
You're welcome, @wuchaotao! Glad I could help clarify how the mask metrics work. If you have any other questions about Ultralytics YOLO models or implementations in the future, feel free to ask.
